### PR TITLE
nix-build: Port to c++

### PR DESCRIPTION
**Pull request by shlevy**
This was a dumb line-for-line rewrite, because nix build/nix run/etc.
will replace it.

### CODE REVIEW

**Comment by vcunat** [comment]
It seems some merge conflicts have originated in the meantime.

**Comment by shlevy** [comment]
Yep, I'll merge + fix conflicts once I get an OK from @edolstra

**Comment by hedning** [comment]
There's a bug, #976, with `nix-shell` due to `$BASH_ENV` being set. Seems like this code also does this. See #1034 for a fix of the perl code.

File: src/nix-build/nix-build.cc [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+        auto args = std::vector<string>{};
+        for (int i = 1; i < argc; ++i)
+            args.push_back(argv[i]);
+        // Heuristic to see if we're invoked as a shebang script, namely, if we
+        // have a single argument, it's the name of an executable file, and it
+        // starts with "#!".
+        if (runEnv && argc > 1 && !std::regex_search(argv[1], std::regex("nix-shell"))) {
+            script = argv[1];
+            if (access(script.c_str(), F_OK) == 0 && access(script.c_str(), X_OK) == 0) {
+                auto SCRIPT = std::ifstream(script);
```
Comment (edolstra) [inline src/nix-build/nix-build.cc]: No capitalized identifiers please.

File: src/nix-build/nix-build.cc [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+                // If we're in a #! script, interpret filenames relative to the
+                // script.
+                if (inShebang && !packages)
+                    expr = absPath(expr, dirOf(script));
+
+                auto instantiateArgs = Strings{"--add-root", drvLink, "--indirect"};
+                for (const auto & arg : instArgs)
+                    instantiateArgs.push_back(arg);
+                instantiateArgs.push_back(expr);
+                auto instOutput = runProgram(settings.nixBinDir + "/nix-instantiate", false, instantiateArgs);
```
Comment (edolstra) [inline src/nix-build/nix-build.cc]: Why call nix-instantiate as a child process when you can do the evaluation in the same process? Without having to do a lot of ugly argument passing etc.

Likewise for the build step.

File: src/nix-build/nix-build.cc [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+#include "store-api.hh"
+#include "globals.hh"
+#include "derivations.hh"
+
+using namespace nix;
+using std::stringstream;
+
+extern char ** environ;
+
+std::vector<string> shellwords(const string & s)
```
Comment (edolstra) [inline src/nix-build/nix-build.cc]: This seems horribly complicated to me. What does it do?

File: src/nix-build/nix-build.cc (hunk 1 above)
Comment (edolstra) [inline src/nix-build/nix-build.cc]: Also I would use `tokenizeString(readFile(script), "\n")` to read the file in lines.

File: src/nix-build/nix-build.cc [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+                        std::smatch match;
+                        if (std::regex_match(line, match, std::regex("^#!\\s*nix-shell (.*)$")))
+                            for (const auto & word : shellwords(match[1].str()))
+                                args.push_back(word);
+                    }
+                }
+            }
+        }
+
+        for (auto n = decltype(args)::size_type{0}; n < args.size(); ++n) {
```
Comment (edolstra) [inline src/nix-build/nix-build.cc]: Ugh, just say `size_t n = 0`.

**Comment by shlevy** [comment]
@edolstra for many of your comments: I figured since `nix build` is in the pipeline, it would be easiest to just dumbly translate the perl because this doesn't have to be maintained for too long (I hope!). Please confirm that you still would rather I remove the child process calls, argument handling, etc.

**Comment by edolstra** [comment]
Yeah that's true. Let's leave it this way for now, we can integrate it will "nix build" later.

**Comment by shlevy** [comment]
OK, which of these comments if any should I still address?

**Comment by edolstra** [comment]
All except the one about using `nix-instantiate` please.

File: src/nix-build/nix-build.cc (hunk 3 above)
Comment (edolstra) [inline src/nix-build/nix-build.cc]: A comment explaining what it does would be good enough.

File: src/nix-build/nix-build.cc (hunk 1 above)
Comment (shlevy) [inline src/nix-build/nix-build.cc]: @edolstra I only read one line here though.

Comment (shlevy) [inline src/nix-build/nix-build.cc]: Ah never mind, sorry

**Comment by shlevy** [comment]
@edolstra Pushed changes

**Comment by edolstra** [comment]
Thanks! Github is showing some merge conflicts though.

**Comment by shlevy** [comment]
Yep, will merge and resolve once it LGTY.

**Comment by edolstra** [comment]
Yes, looks good to me.

**Comment by vcunat** [comment]
:fireworks:

**Pull request merged**